README.md
<a href='https://ko-fi.com/A535IR4' target='_blank'><img height='36' style='border:0px;height:36px;' src='https://az743702.vo.msecnd.net/cdn/kofi4.png?v=f' border='0' alt='Buy Me a Coffee at ko-fi.com' /></a> 
# AutoTable - Table plugin for jsPDF

**Generate PDF tables with Javascript**

This jsPDF plugin aims at making it easy to generate pdf tables either from HTML or directly from Javascript. Check out the [demo](https://simonbengtsson.github.io/jsPDF-AutoTable/) or [examples](https://github.com/simonbengtsson/jsPDF-AutoTable/tree/master/examples).

![sample javascript table pdf](samples.png)

### Installation
Get the library by doing one of these things:
- `npm install jspdf jspdf-autotable`
- Download [jspdf](https://raw.githubusercontent.com/MrRio/jsPDF/master/dist/jspdf.min.js) and [jspdf-autotable](https://raw.githubusercontent.com/simonbengtsson/jsPDF-AutoTable/master/dist/jspdf.plugin.autotable.js) from github
- Use a CDN, for example: [https://unpkg.com/jspdf](https://unpkg.com/jspdf) and [https://unpkg.com/jspdf-autotable](https://unpkg.com/jspdf-autotable)

### Usage example

```html
<table id="my-table"><!-- ... --></table>

<script src="jspdf.min.js"></script>
<script src="jspdf.plugin.autotable.min.js"></script>

<script>
    var doc = new jsPDF();
    // You can use html:
    doc.autoTable({html: '#my-table'});
    
    // Or JavaScript:
    doc.autoTable({
        head: [['Name', 'Email', 'Country']]
        body: [
            ['David', 'david@example.com', 'Sweden'],
            ['Castille', 'castille@example.com', 'Norway'],
            // ...
        ]
    });
    
    doc.save('table.pdf');
</script>
```

Or if using javascript modules and es6:

```js
import jsPDF from 'jspdf';
import 'jspdf-autotable';

const doc = new jsPDF();
doc.autoTable({html: '#my-table'});
doc.save('table.pdf');
```

Checkout more examples in [examples.js](examples) which is also the source code for the [demo](https://simonbengtsson.github.io/jsPDF-AutoTable/) documents.

### API
- `doc.autoTable({ /* options */ })`
- `doc.autoTableSetDefaults({ /* ... */ })` Use for setting default options for all tables in the specific document. Settings and styles will be overridden in the following order `global` < `document` < `table`. Hooks will be added and not overridden.
- `jsPDF.autoTableSetDefaults({ /* ... */ })` Use for setting global defaults which will be applied for all document and tabels.

If you want to know something about the last table that was drawn you can use `doc.lastAutoTable`. It has a `doc.lastAutoTable.finalY` property among other things that has the value of the last printed y coordinate on a page. This can be used to draw text, multiple tables or other content after a table.

### Options
Below is a list of all options supported in the plugin. All of them are used in the [examples](examples).

##### Content options
The only thing required is either the html or body option. If you want more control over the columns you can specify the columns property. It is not needed however and if not set the columns will be automatically computed based on the content of the html content or head, body and foot.

- `html: string|HTMLTableElement` A css selector on an html table element (for example "#table").
- `head: CellDef[][]` For example [['ID', 'Name', 'Country']]
- `body: CellDef[][]` For example [['1', 'Simon', 'Sweden'], ['2', 'Karl', 'Norway']]
- `foot: CellDef[][]` For example [['ID', 'Name', 'Country']]
- `columns: ColumnDef[]` For example [{header: 'ID', dataKey: 'id'}, {header: 'Name', dataKey: 'name'}]. Only use this option if you want more control over the columns. If not specified the columns will be automatically generated based on the content in html or head/body/foot
- `includeHiddenHtml: boolean = false` If hidden html with `display: none` should be included or not when the content comes from an html table

`CellDef: string|{rowSpan: number, colSpan: number, styles: StyleDef}`
Note that cell styles can also be set dynamically with hooks.

`ColumnDef: string|{header?: string, dataKey: string}`
The header property is optional and the values of any content in `head` will be used if not set. Normally it's easier to use the html or head/body/foot style of initiating a table, but columns can be useful if your body content comes directly from an api or if you would like to specify a dataKey on each column to make it more readable to style specific columns in the hooks or columnStyles.

Usage with colspan, rowspan and inline cell styles:

```js
doc.autoTable({
    body: [[{content: 'Text', colSpan: 2, rowSpan: 2, styles: {halign: 'center'}}]]
})
```

##### Styling options

- `theme: 'striped'|'grid'|'plain'|'css' = 'striped'`
- `styles: StyleDef`
- `headStyles: StyleDef`
- `bodyStyles: StyleDef`
- `footStyles: StyleDef`
- `alternateRowStyles: StyleDef`
- `columnStyles: {&columnDataKey: StyleDef}` Note that the columnDataKey is normally the index of the column, but could also be the `dataKey` of a column if content initialized with the columns property

`StyleDef`:
- `font: 'helvetica'|'times'|'courier' = 'helvetica'`
- `fontStyle: 'normal'|'bold'|'italic'|'bolditalic' = 'normal'`
- `overflow: 'linebreak'|'ellipsize'|'visible'|'hidden' = 'normal'`
- `fillColor: Color? = null`
- `textColor: Color? = 20`
- `cellWidth: 'auto'|'wrap'|number = 'auto'`
- `minCellHeight: number = 0`
- `halign: 'left'|'center'|'right' = 'left'`
- `valign: 'top'|'middle'|'bottom' = 'top'`
- `fontSize: number = 10`
- `cellPadding: Padding = 10`
- `lineColor: Color = 10`
- `lineWidth: number = 0` // If 0, no border is drawn

`Color`:
Either false for transparent, rbg array e.g. [255, 0, 0] or gray level e.g 200

`Padding`:
Either a number or object `{top: number, right: number, bottom: number, left: number}`

Styles work similar to css and can be overridden by more specific styles. The overriding order is as follows: 
1. Theme styles
2. `styles`
3. `headStyles`, `bodyStyles` and `footStyles`
4. `alternateRowStyles`
5. `columnStyles`

Styles for specific cells can also be applied using either the hooks (see hooks section above) or the `styles` property on the cell definition object (see content section above).

Example usage of column styles (note that the 0 in the columnStyles below should be dataKey if )

```js
// Example usage with columnStyles, 
doc.autoTable({
    styles: {fillColor: [255, 0, 0]},
    columnStyles: {0: {halign: 'center', fillColor: [0, 255, 0]}}, // Cells in first column centered and green
    margin: {top: 10},
    body: [['Sweden', 'Japan', 'Canada'], ['Norway', 'China', 'USA'], ['Denmark', 'China', 'Mexico']]
})

// Example usage of columns property. Note that America will not be included even though it exist in the body since there is no column specified for it.
doc.autoTable({
    columnStyles: {europe: {halign: 'center'}}, // European countries centered
    body: [{europe: 'Sweden', america: 'Canada', asia: 'China'}, {europe: 'Norway', america: 'Mexico', asia: 'Japan'}],
    columns: [{header: 'Europe', dataKey: 'europe'}, {header: 'Asia', dataKey: 'asia'}]
})
```

##### Other options

- `startY: number = null` Where the table should start to be printed (basically a margin top value only for the first page)
- `margin: Margin = 40`
- `pageBreak: 'auto'|'avoid'|'always'` If set to `avoid` the plugin will only split a table onto multiple pages if table height is larger than page height.
- `rowPageBreak: 'auto'|'avoid' = 'auto'` If set to `avoid` the plugin will only split a row onto multiple pages if row height is larger than page height.
- `tableWidth: 'auto'|'wrap'|number = 'auto'`
- `showHead: 'everyPage'|'firstPage'|'never' = 'everyPage''`
- `showFoot: 'everyPage'|'lastPage'|'never' = 'everyPage''`
- `tableLineWidth: number = 0`
- `tableLineColor: Color = 200` The table line/border color

`Margin`:
Either a number or object `{top: number, right: number, bottom: number, left: number}`

#### Hooks
You can customize the content and styling of the table by using the hooks. See the custom styles example for usage of the hooks.

- `didParseCell: (HookData) => {}` - Called when the plugin finished parsing cell content. Can be used to override content or styles for a specific cell.
- `willDrawCell: (HookData) => {}` - Called before a cell or row is drawn. Can be used to call native jspdf styling functions such as `doc.setTextColor` or change position of text etc before it is drawn.
- `didDrawCell: (HookData) => {}` - Called after a cell has been added to the page. Can be used to draw additional cell content such as images with `doc.addImage`, additional text with `doc.addText` or other jspdf shapes.
- `didDrawPage: (HookData) => {}` - Called after the plugin has finished drawing everything on a page. Can be used to add headers and footers with page numbers or any other content that you want on each page there is an autotable.

All hooks functions get passed an HookData object with information about the state of the table and cell. For example the position on the page, which page it is on etc.

`HookData`:
- `table: Table`
- `pageNumber: number` The page number specific to this table
- `settings: object` Parsed user supplied options 
- `doc` The jsPDF document instance of this table
- `cursor: { x: number, y: number }` To draw each table this plugin keeps a cursor state where the next cell/row should be drawn. You can assign new values to this cursor to dynamically change how the cells and rows are drawn.

For cell hooks these properties are also passed:
- `cell: Cell`
- `row: Row`
- `column: Column`
- `section: 'head'|'body'|'foot'`

To see what is included in the `Table`, `Row`, `Column` and `Cell` types, either log them to the console or take a look at `src/models.ts`

```js
// Example with an image drawn in each cell in the first column
doc.autoTable({
   didDrawCell: data => {
      if (data.section === 'body' && data.column.index === 0) {
         var base64Img = "data:image/jpeg;base64,iVBORw0KGgoAAAANS...";
         doc.addImage(base64Img, 'JPEG', data.cell.x + 2, data.cell.y + 2, 10, 10);
      }
   }
})
```

### Contributions
Contributions are always welcome, especially on open issues. If you have something major you want to add or change, please post an issue about it first to discuss it further. The workflow for contributing would be something like this:

- Start watcher with `npm start`
- Make code changes
- Make sure all examples works
- Commit and submit pull request

### Release workflow (write access required)
- Test and commit code changes 
- `npm version <semver|major|minor|patch>`
- Verify examples
- `npm run deploy`

### Release prerelease
- `npm version prerelease`
- `git push && git push --tags && npm publish --tag alpha`


CHANGELOG.md
# Known issues 3.0 alpha (to be fixed before release)
- Go over popular stack overflow and codepens
- Proofread docs and config
- Test IE9

### 3.1
- Column styles with the ColumnDef
- Export API functions as well as adding them to jsPDF prototype
- Look into if tree shaking is possible
- Only print head if first row of body fits on page
- Additional and updated themes?
- Easier way to add image to cells
- Easier way to add links to cells
- Update support for custom fonts (and utf-8)?
- Support dividing wide tables horizontally on multiple pages
- Rewrite docs and examples
- Handle column widths less than 10 and larger than page width
- Improved column width calculation (decrease wide columns more)
- More advanced column props (e.g. with support for footer and header groups?)
- Refactor with typescript support and improved options etc

### Content arguments
- Most seem to be using html initiation and head,body,foot is no new concept to learn. It might be some trade-offs for more experienced users and in terms of logic. But a columns field can be used which would be same as v2

# Changelog

Migrating to version 3.0 should be rather painless as most changes are backwards compatible. The lib was pretty much rewritten however so open an issue if you encounter any problems.

BREAKING: The hooks API has been simplified. Note that colspans and rowspans now is supported so you can use that instead of the old drawCell hack.
- Removed hooks: createdCell, createdHeaderCell, drawHeaderCell, drawCell, drawRow, drawHeaderRow and addPageContent
- Added hooks: didParseCell, willDrawCell, didDrawCell and didDrawPage

All other changes should be backwards compatible, but it's a rewrite of the plugin so file an issue if you encounter something that did not work as expected after upgrading.

### 3.0
- Added native rowspan and colspan support
- Added multiple header rows support
- Added footer rows support
- Added html initialization with `html: string|HTMLTableElement`. If set to string it should be a css selector pointing to a table element.
- Added `useCss: boolean` for using some basic css styles when table initialized from html
- Added `includeHiddenHtml: boolean` option
- Added `showFooter` option (similar to showHeader)
- Added `footerStyles` option (similar to headerStyles)
- Added `rowPageBreak` option
- Added automatic startY when multiple tables are used
- Changed initialization to `doc.autoTable({head: ..., body: ... foot: ..., columns: ...})`
- Changed getting last autoTable instance from `doc.autoTable.previous` to `doc.lastAutoTable`
- Changed default overflow method to linebreak (previously it was ellipsize)
- Changed `rowHeight` style to `cellHeight`
- Changed `columnWidth` style to `cellWidth`
- Changed and simplified hooks to `didParseCell, willDrawCell, didDrawCell, didDrawPage
- Deprecated the old initialization `doc.autoTable(columns, data, options)`
- Deprecated `autoTableHtmlToJson()` in favour of new html option
- Deprecated old way of getting previous autoTable instance `doc.autoTable.previous`
- BREAKING: Old hooks with new ones (See above)
- BREAKING: No longer supports 0.9.x of jsPDF
- BREAKING: Now targets ES5 (no longer supports <=IE8)


src/config.js
/**
 * Ratio between font size and font height. The number comes from jspdf's source code
 */
export let FONT_ROW_RATIO = 1.15;
import {Table} from './models.js';

let jspdfInstance = null;
let userStyles = null;
let settingsObject = null; // Default options merged with user options
let table = null;
let globalAddPageContent = function() {}; // Override with doc.autoTableAddPageContent

/**
 * Styles for the themes (overriding the default styles)
 */
export let getTheme = function(name) {
    let scaleFactor = Config.getJspdfInstance().internal.scaleFactor;
    let themes = {
        'striped': {
            table: {fillColor: 255, textColor: 80, fontStyle: 'normal'},
            header: {textColor: 255, fillColor: [41, 128, 185], rowHeight: 23 / scaleFactor, fontStyle: 'bold'},
            body: {},
            alternateRow: {fillColor: 245}
        },
        'grid': {
            table: {fillColor: 255, textColor: 80, fontStyle: 'normal', lineWidth: 0.1},
            header: {textColor: 255, fillColor: [26, 188, 156], rowHeight: 23 / scaleFactor, fontStyle: 'bold', lineWidth: 0},
            body: {},
            alternateRow: {}
        },
        'plain': {
            header: {fontStyle: 'bold'}
        }
    };
    return themes[name];
};

function getDefaults() {
    let scaleFactor = Config.getJspdfInstance().internal.scaleFactor;
    return {
        // Styling
        theme: 'striped', // 'striped', 'grid' or 'plain'
        styles: {},
        headerStyles: {},
        bodyStyles: {},
        alternateRowStyles: {},
        columnStyles: {},

        // Properties
        startY: false, // false indicates the margin.top value
        margin: 40 / scaleFactor,
        pageBreak: 'auto', // 'auto', 'avoid', 'always'
        tableWidth: 'auto', // 'auto'|'wrap'|number (takes precedence over columnWidth style if conflict)
        showHeader: 'always', // 'always', 'once', 'never',

        // Hooks
        createdHeaderCell: function (cell, data) {},
        createdCell: function (cell, data) {},
        drawHeaderRow: function (row, data) {},
        drawRow: function (row, data) {},
        drawHeaderCell: function (cell, data) {},
        drawCell: function (cell, data) {},
        addPageContent: function (data) {}
    }
}

// Base style for all themes
function defaultStyles() {
    let scaleFactor = Config.getJspdfInstance().internal.scaleFactor;
    return {
        font: "helvetica", // helvetica, times, courier
        lineColor: 200,
        fontStyle: 'normal', // normal, bold, italic, bolditalic
        overflow: 'ellipsize', // visible, hidden, ellipsize or linebreak
        fillColor: false, // Either false for transparant, rbg array e.g. [255, 255, 255] or gray level e.g 200
        textColor: 20,
        halign: 'left', // left, center, right
        valign: 'top', // top, middle, bottom
        fontSize: 10,
        cellPadding: 5 / scaleFactor,
        lineWidth: 0 / scaleFactor,
        rowHeight: 20 / scaleFactor,
        columnWidth: 'auto'
    }
}

export class Config {
    
    static setJspdfInstance(instance) {
        jspdfInstance = instance;
        userStyles = {
            textColor: 30, // Setting text color to dark gray as it can't be obtained from jsPDF
            fontSize: jspdfInstance.internal.getFontSize(),
            fontStyle: jspdfInstance.internal.getFont().fontStyle
        };
    }
    
    static getJspdfInstance() {
        return jspdfInstance;
    }
    
    // Styles before autotable was called
    static getUserStyles() {
        return userStyles;
    }
    
    static createTable() {
        table = new Table();
    }
    
    static setPageContentHook(hook) {
        globalAddPageContent = hook;
    }
    
    static callPageContentHook(data) {
        if (typeof globalAddPageContent === 'function') {
            globalAddPageContent(data);
        }
    }
    
    static tableInstance() {
        return table;
    }
    
    static settings() {
        return settingsObject;
    }

    static hooksData(additionalData) {
        return Object.assign({
            pageCount: table.pageCount,
            settings: settingsObject,
            table: table,
            doc: jspdfInstance,
            cursor: jspdfInstance.autoTableCursor
        }, additionalData || {});
    }

    static initSettings(userOptions) {
        let settings = Object.assign({}, getDefaults(), userOptions);

        // Options
        if (typeof settings.extendWidth !== 'undefined') {
            settings.tableWidth = settings.extendWidth ? 'auto' : 'wrap';
            console.error("Use of deprecated option: extendWidth, use tableWidth instead.");
        }
        if (typeof settings.margins !== 'undefined') {
            if (typeof settings.margin === 'undefined') settings.margin = settings.margins;
            console.error("Use of deprecated option: margins, use margin instead.");
        }
        if (typeof settings.afterPageContent !== 'undefined' || typeof settings.beforePageContent !== 'undefined' || typeof settings.afterPageAdd !== 'undefined') {
            console.error("The afterPageContent, beforePageContent and afterPageAdd hooks are deprecated. Use addPageContent instead");
            if (typeof userOptions.addPageContent === 'undefined') {
                settings.addPageContent = function(data) {
                    Config.applyStyles(Config.getUserStyles());
                    if (settings.beforePageContent) settings.beforePageContent(data);
                    Config.applyStyles(Config.getUserStyles());
                    if (settings.afterPageContent) settings.afterPageContent(data);
                    Config.applyStyles(Config.getUserStyles());

                    if (settings.afterPageAdd && data.pageCount > 1) {
                        data.afterPageAdd(data);
                    }
                    Config.applyStyles(Config.getUserStyles());
                }
            }
        }

        [['padding', 'cellPadding'], ['lineHeight', 'rowHeight'], 'fontSize', 'overflow'].forEach(function (o) {
            let deprecatedOption = typeof o === 'string' ? o : o[0];
            let style = typeof o === 'string' ? o : o[1];
            if (typeof settings[deprecatedOption] !== 'undefined') {
                if (typeof settings.styles[style] === 'undefined') {
                    settings.styles[style] = settings[deprecatedOption];
                }
                console.error("Use of deprecated option: " + deprecatedOption + ", use the style " + style + " instead.");
            }
        });
        
        settings.margin = Config.marginOrPadding(settings.margin, 40);

        settingsObject = settings;
    }
    
    static marginOrPadding(value, defaultVal) {
        let newValue = {};
        ['top', 'right', 'bottom', 'left'].forEach(function (side, i) {
            let k = Config.getJspdfInstance().internal.scaleFactor;
            newValue[side] = defaultVal / k;
            if (typeof value === 'number') {
                newValue[side] = value;
            } else if (Array.isArray(value) && typeof value[i] === 'number') {
                newValue[side] = value[i];
            } else if (typeof value === 'object') {
                if (typeof value[side] === 'number') {
                    newValue[side] = value[side];
                } else if ((side === 'right' || side === 'left') && typeof value['horizontal'] === 'number') {
                    newValue[side] = value['horizontal'];
                } else if ((side === 'top' || side === 'bottom') && typeof value['vertical'] === 'number') {
                    newValue[side] = value['vertical'];
                }
            }
        });
        return newValue;
    }

    static styles(styles) {
        let defStyles = defaultStyles();
        styles.cellPadding = Config.marginOrPadding(styles.cellPadding, defStyles.cellPadding);
        return Object.assign({}, defStyles, ...styles);
    }

    static applyStyles(styles) {
        let doc = Config.getJspdfInstance();
        let styleModifiers = {
            fillColor: doc.setFillColor,
            textColor: doc.setTextColor,
            fontStyle: doc.setFontStyle,
            lineColor: doc.setDrawColor,
            lineWidth: doc.setLineWidth,
            font: doc.setFont,
            fontSize: doc.setFontSize
        };
        Object.keys(styleModifiers).forEach(function (name) {
            let style = styles[name];
            let modifier = styleModifiers[name];
            if (typeof style !== 'undefined') {
                if (Array.isArray(style)) {
                    modifier.apply(this, style);
                } else {
                    modifier(style);
                }
            }
        });
    } 
}